cirq-core/cirq/value/condition.py

```diff
@@ -0,0 +1,168 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import abc
+import dataclasses
+import re
+from typing import Dict, Mapping, Sequence, Tuple, TYPE_CHECKING
+
+import sympy
+
+from cirq.protocols import json_serialization
+from cirq.value import digits, measurement_key
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class Condition(abc.ABC):
+    """A classical control condition that can gate an operation."""
+
+    @property
+    @abc.abstractmethod
+    def keys(self) -> Tuple['cirq.MeasurementKey', ...]:
+        """Gets the control keys."""
+
+    @abc.abstractmethod
+    def with_keys(self, *keys: 'cirq.MeasurementKey'):
+        """Replaces the control keys."""
+
+    @abc.abstractmethod
+    def resolve(self, measurements: Mapping[str, Sequence[int]]) -> bool:
+        """Resolves the condition based on the measurements."""
+
+    @property
+    @abc.abstractmethod
+    def qasm(self):
+        """Returns the qasm of this condition."""
+
+
+@dataclasses.dataclass(frozen=True)
+class KeyCondition(Condition):
+    """A classical control condition based on a single measurement key.
+
+    This condition resolves to True iff the measurement key is non-zero at the
+    time of resolution.
+    """
+
+    key: 'cirq.MeasurementKey'
+
+    @property
+    def keys(self):
+        return (self.key,)
+
+    def with_keys(self, *keys: 'cirq.MeasurementKey'):
+        assert len(keys) == 1
+        return KeyCondition(keys[0])
+
+    def __str__(self):
```

Comment:
Cirq classes require a `__repr__` method for which `obj = eval(repr(obj))`. This is also usually the source of the `Class.repr` file in the `json_test_data` directory.

Same for other serializable classes in this PR.